clippy_lints/src/casts/unnecessary_cast.rs

```diff
@@ -80,6 +82,17 @@ pub(super) fn check<'tcx>(
                 }
             },
         }
+    } else if cast_from.kind() == cast_to.kind() && !in_external_macro(cx.sess(), expr.span) {
+        span_lint_and_sugg(
+            cx,
+            UNNECESSARY_CAST,
+            expr.span,
+            &format!("casting to the same type is unnecessary (`{cast_from}` -> `{cast_to}`)"),
+            "try",
+            cast_str,
+            Applicability::MachineApplicable,
+        );
+        return true;
```

Comment:
Looks like it should already be caught by https://github.com/rust-lang/rust-clippy/blob/783ee635559a522a7a46b95419ccc536c6a28f97/clippy_lints/src/casts/unnecessary_cast.rs#L71
no?

Reply:
It's under a condition so it will only work for numeric literal.

Reply:
right, you could refactor it by turning your `if else` into a `if` and remove the afore mentioned line then I think

Reply:
I've tried it when I implement it, but the control flow is too complicated to do that. At least, I will try to extract a common function `lint_same_type`.

Reply:
Updated